src/react-render-perf/lesson-01/README.md

```diff
@@ -0,0 +1,137 @@
+the component itself is expensive to render (e.g. the `MathJax` component in 
+webapp) or it renders a lot of descedent components.
+
+Memoization works by saving the rendered output of the component based on the
+props that are being passed in. Often times props will appear to have changed
+when their actual values haven't. In JavaScript, two objects with the same 
+properties are considered different objects. Similarly, two functions with the
+same implementation are considered differen objects.
+
+In order for memoization to have the desired benefit, we want don't want the
+component to rerender if there are only superficial changes to props.
+
+`React.memo(Component, arePropsEqual?)` by default does a shallow comparison of
+props and as such isn't able to determine when a prop that's on object or function
+is the same or not. We can pass a custom `arePropsEqual` function to override
+that behavior. To keep things simple we use a third-party library called
+`react-fast-compare` which provides a function that does a deep comparison of
+objects.
+
+```ts
+import arePropsEqual from "react-fast-compare";
+
+type Props = {
+    user: {name: string, id: string},
+    onClick: () => void,
+}
+
+const ChildComponent = (props: Props) => {
+    // ...
+}
+
+export default React.memo(ChildComponent, arePropsEqual);
+```
+
+There is a bit of a gotcha here when it comes to props that are functions.
+`react-fast-compare` cannot check if two functions are the same. Imagine the
+following scenario:
+
+```ts
+import ChildComponent from "./child-component";
+
+type Props = {
+    user: {name: string, id: string},
+};
+
+const ParentComponent = (props: Props) => {
+    const result = useQuery(QUERY);
+
+    const handleClick = () => {
+        if (result.data) {
+            // do something with the data
+        }
+    };
+
+    return <ChildComponent user={user} onClick={handleClick}>
+}
+```
+
+Each time `ParentComponent` renders, a new copy of `handleClick` will be created
+even if the `result` from `useQuery` isn't ready yet. We only want this function
+to treated as a new function when `result` changes. React provides a hook called
+`useCallback` which does exactly that by memoizing the function.
+
+```ts
+import ChildComponent from "./my-component";
+
+type Props = {
+    user: {name: string, id: string},
+};
+
+const ParentComponent = (props: Props) => {
+    const result = useQuery(QUERY);
+
+    const handleClick = React.useCallback(() => {
+        if (result.data) {
+            // do something with the data
+        }
+    }, [result]);
+
+    return <ChildComponent user={user} onClick={handleClick}>
+}
+```
+
+If the `ParentComponent` is a class-based component, there is no need to memoize
+function props that are pre-bound methods. This is because the method never changes
+for the component instance. If the prop is an inline function though, it should be
+convered to a pre-bound method.
```

Comment:
"inline function" is confusing to me here -- do you mean a regular class [method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Method_definitions)? And what you call "pre-bound" method makes some sense as a name, but I haven't heard it used before (although I might just not be in the loop). From a technical standpoint, `x = () =>` in a class is a ["class field"](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Public_class_fields) (though I've also heard them called "attributes" or "properties"), which is distinct from a [method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Method_definitions), which lives on the class's prototype.

Reply:
Inline like `onClick={() => { ... }}`.  I'll add an example so that this part is clearer.